Added missing algorithm examples in documentation.

digu-007 opened the pull request:
This PR fixes  issue #666.

I have added all the files as stated in the issue, with proper outputs and comments.

Kindly, please review.

mloskot commented:
@digu-007 I suggest you to have a look at the ongoing review of another example in PR #665 and auto-review your PR addressing the comments posted by @vissarion to the #665, if applicable of coruse.
Chances are you will clear some of potential issues.

barendgehrels commented:
Looks promising - but they are not called yet. They need to be referred to from source code.
See indeed also the comments of #665 and please look in one of the existing examples to see how they are referred to

digu-007 commented:
> Looks promising - but they are not called yet. They need to be referred to from source code.
> See indeed also the comments of #665 and please look in one of the existing examples to see how they are referred to

Referred to from source code means that you need me to add exact header files of the algorithms?
I will work on the doxygen commands as well.

vissarion commented:
>Referred to from source code means that you need me to add exact header files of the algorithms?
I will work on the doxygen commands as well.

No need to add extra header files, just inject `doxygen` code in them. 

See for example `densify` algorithm:
https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/algorithms/densify.hpp#L401

And you have to inject `doxygen` code in you example as well.
See:
https://github.com/boostorg/geometry/blob/develop/doc/src/examples/algorithms/densify.cpp#L11

vissarion commented:
Adding images for the geometries in your example will give an extra value on the documentation and help the reader even more.

digu-007 commented:
> Adding images for the geometries in your example will give an extra value on the documentation and help the reader even more.

Ok sure, working on it.

digu-007 commented:
> Adding images for the geometries in your example will give an extra value on the documentation and help the reader even more.

Can you please tell me the importance of `/*<-*/ create_svg("union.svg", green, blue, output); /*->*/`?
I have made the png file for the required image, and is writing `[$img/algorithms/covered_by.png]` at the bottom of output enough?

vissarion commented:
> > Adding images for the geometries in your example will give an extra value on the documentation and help the reader even more.
> 
> Can you please tell me the importance of `/*<-*/ create_svg("union.svg", green, blue, output); /*->*/`?
> I have made the png file for the required image, and is writing `[$img/algorithms/covered_by.png]` at the bottom of output enough?

this is code (commented out to not appear in the example) to create the `svg` file  for the example and is used for future reference. AFAIU, the `png` displayed in the documentation is generated by manually converting the generated `svg` to `png`  in some application (e.g. `inkscape`). So creating the `png` and storing it to `html/img/algorithms/` should be fine.

digu-007 commented:
> > > Adding images for the geometries in your example will give an extra value on the documentation and help the reader even more.
> > 
> > 
> > Can you please tell me the importance of `/*<-*/ create_svg("union.svg", green, blue, output); /*->*/`?
> > I have made the png file for the required image, and is writing `[$img/algorithms/covered_by.png]` at the bottom of output enough?
> 
> this is code (commented out to not appear in the example) to create the `svg` file for the example and is used for future reference. AFAIU, the `png` displayed in the documentation is generated by manually converting the generated `svg` to `png` in some application (e.g. `inkscape`).
> 
> So creating the `png`, adding something like `[$img/algorithms/covered_by.png]` and storing the `png` to `html/img/algorithms/` should be fine.
> 
> It is recomened to generate the documentation on your side and test everything is OK.

I added images and example files to imports.qbk [here](https://github.com/boostorg/geometry/pull/667/commits/f6e32351ff7379d584c6364e6495eb8ecbe0d1ac). Can you please review it.

I'm getting bugs while generating the documentation, can you help me with that?

mloskot commented:
> I'm getting bugs while generating the documentation, can you help me with that?

You need to be (very) specific:
- how do you build the docs exactly
- what commands you run exactly
- what errors you get exactly

Mind you, crystal spheres are not always handy with their guessing magic.

digu-007 commented:
> > I'm getting bugs while generating the documentation, can you help me with that?
> 
> You need to be (very) specific:
> 
> * how do you build the docs exactly
> * what commands you run exactly
> * what errors you get exactly
> 
> Mind you, crystal spheres are not always handy with their guessing magic.

I followed [Generating-and-Improving-Documentation](url), and while running `make_qbk.py` I got following error.
![error](https://user-images.githubusercontent.com/31317843/74243401-b41a0200-4d05-11ea-8de0-f7d494f9f204.png)

digu-007 commented:
> > I'm getting bugs while generating the documentation, can you help me with that?
> 
> You need to be (very) specific:
> 
> * how do you build the docs exactly
> * what commands you run exactly
> * what errors you get exactly
> 
> Mind you, crystal spheres are not always handy with their guessing magic.

I followed [Generating-and-Improving-Documentation](https://github.com/boostorg/geometry/wiki/Generating-and-Improving-Documentation), and while executing `make_qbk.py`, I got the following error.
![error](https://user-images.githubusercontent.com/31317843/74243679-30144a00-4d06-11ea-85b5-f931192c7ba7.png)

mloskot commented:
That Wiki page say:

> Compile doxygen_xml2qbk tool shipped with Boost.Geometry....

which I presume you have not attempted.
That is why you are getting the `doxygen_xml2qbk` not found error.

I also think you probably should use the current `develop `branch, not the Boost 1.72 archive, especially, that you are submitting PRs which can only be submitted against the `develop` branch.

digu-007 commented:
> That Wiki page say:
> 
> > Compile doxygen_xml2qbk tool shipped with Boost.Geometry....
> 
> which I presume you have not attempted.
> That is why you are getting the `doxygen_xml2qbk` not found error.
> 
> I also think you probably should use the current `develop `branch, not the Boost 1.72 archive, especially, that you are submitting PRs which can only be submitted against the `develop` branch.

I am using `develop` branch only for PRs and not the Boost 1.72, it's just that I was building the documentation using Boost 1.72 just to check how to build, but now I won't use it.

Yes I forgot about `doxygen_xml2qbk`.

I installed Doxygen and QuickBook with the help of [this](https://github.com/boostorg/geometry/blob/develop/doc/readme.txt) but while running `b2`, I'm getting following error:
![this](https://user-images.githubusercontent.com/31317843/74249133-ae74ea00-4d0e-11ea-8e03-884cef92a27a.png)

Also can you please verify [latest commit](https://github.com/boostorg/geometry/pull/667/commits/f6e32351ff7379d584c6364e6495eb8ecbe0d1ac). I made the necessary changes to `imports.qbk` and added images for better understanding of examples.

mloskot commented:
> I was building the documentation using Boost 1.72 just to check how to build

You can build documentation from git repo. I'd simply work in the cloned Boost superproject as it is the common way any Boost library is developed.

> I'm getting following error ...

I suspect you just cloned the `boostorg/geometry` repository.

You need to **clone the whole Boost super-project** as it is explained in

- this Geometry's https://github.com/boostorg/geometry/wiki/Contribution-Tutorial
- general Boost development getting started tutorial https://github.com/boostorg/boost/wiki/Getting-Started%3A-Overview
- in documentation of other Boost libraries - the initial clone-build-test workflow is almost exactly the same for all Boost libraries e.g. GIL, https://github.com/boostorg/gil/blob/develop/CONTRIBUTING.md#getting-started-with-git-workflow 

> Also can you please verify latest commit. 

First, let's get your documentation build to working state.

digu-007 commented:
> You need to **clone the whole Boost super-project** as it is explained in
> 
> * this Geometry's https://github.com/boostorg/geometry/wiki/Contribution-Tutorial

I cloned the whole Boost super-project, switched to develop branch and ran `.\bootstrap.sh`. But `.\b2 headers` is giving error `zsh: command not found: .b2`

mloskot commented:
@digu-007 What shell/console/command line are you familiar with, Windows, Unix or both? Are you familiar with the difference between `.\program` vs `./program` conventions?

I suspect you are using Linux. I'd say, it should be quite plain obvious for you to run `./b2`.

The Wiki pages try to do its best, present `git` and `b2` commands a reader needs to run, but they do not cover all conventions for all command lines. So, a reader needs to apply environments-specific adjustments, obviously.

digu-007 commented:
> The Wiki pages try to do its best, present `git` and `b2` commands a reader needs to run, but they 

I did `./b2 headers` and got output `...found 1171 targets...`. I did `export PATH=$PATH:$HOME/boost` to add Boost directory to PATH but `b2 test` is not working.

```
/home/digu_j/boost/tools/build/src/tools/features/warnings-feature.jam:27: in load
ERROR: rule "[[bbv2.builtin.features.warnings]]`warnings`::" unknown in module "warnings-feature".
/home/digu_j/boost/tools/build/src/kernel/modules.jam:295: in import
/home/digu_j/boost/tools/build/src/tools/features/__init_features__.jam:22: in load
/home/digu_j/boost/tools/build/src/kernel/modules.jam:295: in import
/home/digu_j/boost/tools/build/src/tools/builtin.jam:16: in load
/home/digu_j/boost/tools/build/src/kernel/modules.jam:295: in import
/home/digu_j/boost/tools/build/src/build-system.jam:12: in load
/home/digu_j/boost/tools/build/src/kernel/modules.jam:295: in import
/home/digu_j/boost/tools/build/src/kernel/bootstrap.jam:139: in boost-build
/home/digu_j/boost/boost-build.jam:17: in module scope
```

mloskot commented:
> I did ./b2 headers and got output ...found 1171 targets...

Good, that seems like you've got all headers deployed/symlinked in `boost/` sub-directory of the cloned Boost repository.

> I did export `PATH=$PATH:$HOME/boost` to add Boost directory to `PATH`

That is not strictly necessary. You can just run `b2` specifying absolute or relative path.

Try to replicate the commands I run below:

```console
$ pwd
/mnt/d/boost.wsl

$ ls
CMakeLists.txt  Jamroot          README.md     b2     boost-build.jam  boost.png     bootstrap.bat  bootstrap.sh  doc        index.html  more                rst.css  tools
INSTALL         LICENSE_1_0.txt  appveyor.yml  boost  boost.css        boostcpp.jam  bootstrap.log  dist          index.htm  libs        project-config.jam  status

$ ./b2 headers
Performing configuration checks

    - default address-model    : 64-bit (cached)
    - default architecture     : x86 (cached)
    - symlinks supported       : yes (cached)
...found 1171 targets...

$ ls ./boost
accumulators             concept                      filesystem.hpp

$ cd libs/geometry/

$ ../../b2 test
Performing configuration checks

    - default address-model    : 64-bit
    - default architecture     : x86
    - symlinks supported       : yes
...patience...
...patience...
...patience...
...patience...
...found 11821 targets...
...updating 2757 targets...
...on 100th target...
...on 200th target...
gcc.compile.c++ ../../bin.v2/libs/geometry/test/util/util_math_abs.test/gcc-9/debug/threading-multi/visibility-hidden/math_abs.o
gcc.compile.c++ ../../bin.v2/libs/geometry/test/util/util_math_sqrt.test/gcc-9/debug/threading-multi/visibility-hidden/math_sqrt.o
gcc.compile.c++ ../../bin.v2/libs/geometry/test/util/util_calculation_type.test/gcc-9/debug/threading-multi/visibility-hidden/calculation_type.o
gcc.link ../../bin.v2/libs/geometry/test/util/util_math_abs.test/gcc-9/debug/threading-multi/visibility-hidden/util_math_abs
testing.capture-output ../../bin.v2/libs/geometry/test/util/util_math_abs.test/gcc-9/debug/threading-multi/visibility-hidden/util_math_abs.run
**passed** ../../bin.v2/libs/geometry/test/util/util_math_abs.test/gcc-9/debug/threading-multi/visibility-hidden/util_math_abs.test
...
```

digu-007 commented:
> $ ../../b2 test
These tests are running for more than two hours in my system, is there any way to up speed this process?

mloskot commented:
Buy more cores, e.g. 16 or more ;) 

It's thousands of programs to compile and link. However, just kill it. You don't have to run all those tests for your purpose of docs. You run that command and let a bunch of tests build as verification you have basic setup ready and b2 command works.

digu-007 commented:
> Buy more cores, e.g. 16 or more ;)

Maybe ;)

Yes, b2 works and I guess the setup is ready. I will run the qbk files tomorrow to generate the changes that I made and check for any errors.

digu-007 commented:
This time I ran `./b2 libs/geometry/doc/src/docutils/tools/doxygen_xml2qbk` successfully and compiled tool got placed in `boost/dist/bin`. I exported this to `PATH` and then I ran `python make_qbk.py` but it is still giving me the similar error.
```
Boost.Geometry is making .qbk files in /home/digu_j/boost/libs/geometry/doc
Boost.Geometry is cleaning Doxygen files in /home/digu_j/boost/libs/geometry/doc/doxy/doxygen_output/xml/
performance:-1: warning: multiple use of section label 'performance', (first occurrence: /home/digu_j/boost/libs/geometry/doc/doxy/doxygen_input/pages/doxygen_pages.hpp, line 152)
Warning: template parameter Point has no type in franklin
Warning: template parameter PointOfSegment has no type in franklin
Warning: template parameter Point has no type in crossings_multiply
Warning: template parameter PointOfSegment has no type in crossings_multiply
Parameter not found: width_height
/home/digu_j/boost/boost/geometry/index/equal_to.hpp:165: warning: The following parameters of boost::geometry::index::detail::equal_to::operator()(Value const &l, Value const &r, Strategy const &strategy) const are not documented:
  parameter 'strategy'
/home/digu_j/boost/boost/geometry/index/equal_to.hpp:225: warning: The following parameters of boost::geometry::index::detail::equal_to< boost::tuple< T0, T1, T2, T3, T4, T5, T6, T7, T8, T9 >, false >::operator()(value_type const &l, value_type const &r, Strategy const &strategy) const are not documented:
  parameter 'strategy'
/home/digu_j/boost/boost/geometry/index/equal_to.hpp:194: warning: The following parameters of boost::geometry::index::detail::equal_to< std::pair< T1, T2 >, false >::operator()(std::pair< T1, T2 > const &l, std::pair< T1, T2 > const &r, Strategy const &strategy) const are not documented:
  parameter 'strategy'
/home/digu_j/boost/boost/geometry/index/equal_to.hpp:288: warning: The following parameters of boost::geometry::index::detail::equal_to< std::tuple< Args... >, false >::operator()(value_type const &l, value_type const &r, Strategy const &strategy) const are not documented:
  parameter 'strategy'
/home/digu_j/boost/boost/geometry/index/equal_to.hpp:165: warning: The following parameters of boost::geometry::index::detail::equal_to::operator()(Value const &l, Value const &r, Strategy const &strategy) const are not documented:
  parameter 'strategy'
/home/digu_j/boost/boost/geometry/index/equal_to.hpp:225: warning: The following parameters of boost::geometry::index::detail::equal_to< boost::tuple< T0, T1, T2, T3, T4, T5, T6, T7, T8, T9 >, false >::operator()(value_type const &l, value_type const &r, Strategy const &strategy) const are not documented:
  parameter 'strategy'
/home/digu_j/boost/boost/geometry/index/equal_to.hpp:194: warning: The following parameters of boost::geometry::index::detail::equal_to< std::pair< T1, T2 >, false >::operator()(std::pair< T1, T2 > const &l, std::pair< T1, T2 > const &r, Strategy const &strategy) const are not documented:
  parameter 'strategy'
/home/digu_j/boost/boost/geometry/index/equal_to.hpp:288: warning: The following parameters of boost::geometry::index::detail::equal_to< std::tuple< Args... >, false >::operator()(value_type const &l, value_type const &r, Strategy const &strategy) const are not documented:
  parameter 'strategy'
/home/digu_j/boost/tools/build/src/tools/features/warnings-feature.jam:27: in load
ERROR: rule "[[bbv2.builtin.features.warnings]]`warnings`::" unknown in module "warnings-feature".
/home/digu_j/boost/tools/build/src/kernel/modules.jam:295: in import
/home/digu_j/boost/tools/build/src/tools/features/__init_features__.jam:22: in load
/home/digu_j/boost/tools/build/src/kernel/modules.jam:295: in import
/home/digu_j/boost/tools/build/src/tools/builtin.jam:16: in load
/home/digu_j/boost/tools/build/src/kernel/modules.jam:295: in import
/home/digu_j/boost/tools/build/src/build-system.jam:12: in load
/home/digu_j/boost/tools/build/src/kernel/modules.jam:295: in import
/home/digu_j/boost/tools/build/src/kernel/bootstrap.jam:139: in boost-build
/home/digu_j/boost/boost-build.jam:17: in module scope
Traceback (most recent call last):
  File "make_qbk.py", line 219, in <module>
    run_command("b2")
  File "make_qbk.py", line 22, in run_command
    
Exception: Error running b2
```

mloskot commented:
>  then I ran python make_qbk.py 

Please, again, be specific, show all relevant details of what you are doing.

You show the command output, only.
You need to show what command are you running **exactly**:
- in which folder you are (where do you `cd`)
- what arguments you pass

Since the output does not show any error like `b2: not found`, I assume you are still having your `b2` in `PATH`, correct? Is it `b2` built with the initial `./bootstrap.sh` step? That is, `b2` build in the cloned repo of Boost.

In order to run `make_qbk.py` you do need `b2` in `PATH`. I just need to confirm you have the right one in `PATH` and that you are **not** using `b2` installed from any Ubuntu or Debian or other packages of your distro.

digu-007 commented:
> I will run the qbk files tomorrow to generate the changes that I made and check for any errors.

I ran `libs/geometry/doc/src/docutils/tools/doxygen_xml2qbk` successfully and compiled tool was placed in `boost/dist/bin`. I ran `python make_qbk.py` after exporting `boost/dist/bin` to `PATH` and got output `...updated 3 targets...` at the end and no errors.
So this means documentation is built successfully and I can test my modifications now?

PS: You replied to my yesterday's comment.

mloskot commented:
> So this means documentation is built successfully and I can test my modifications now?

Good. Yes, I think so.

> PS: You replied to my yesterday's comment.

Sorry, I read comments via e-mail and apparently missed further updates.

digu-007 commented:
> Good. Yes, I think so.

Now to check my updates, I would have to run all these steps again and check the html file for output right? ( Starting from `libs/geometry/doc/src/docutils/tools/doxygen_xml2qbk`)

barendgehrels commented:
Hi @digu-007 thanks for your efforts! Yes, it starts to look good now.

The examples themselves look complete - but I think there is still not yet referred to. It is a complex system at first (and even after a while too). The best you can do is, I think, look how it is done with other examples.

For example for the buffer_join_round, the headerfile contains:
\qbk{
[heading Example]
[buffer_join_round]
[heading Output]
[$img/strategies/buffer_join_round.png]

}
And this will insert your example to the algorithm description.

barendgehrels commented:
> 
> 
> > Good. Yes, I think so.
> 
> Now to check my updates, I would have to run all these steps again and check the html file for output right? ( Starting from `libs/geometry/doc/src/docutils/tools/doxygen_xml2qbk`)

Exactly (but the steps compiling the tools you can probably omit now)

digu-007 commented:
Where could I load the html page of an algorithm from, after editing and running `python make_qbk.py`?

`/boost/libs/geometry/doc/html` is showing only `img  index.html  svg_simplify_country.png  svg_simplify_road.png`

mloskot commented:
@digu-007 
- Have you followed the Quickbook installation as per `libs/geometry/doc/readme.txt`?
- Have you created `$HOME/user-config.jam` with the `boostbook` and other entries as explained on http://www.boost.org/doc/libs/release/doc/html/quickbook/install.html?

Once you have that configuration in place, you will see 4 targets built:

```console
$ cd libs/geometry/doc
$ python3 make_qbk.py
...
Performing configuration checks

    - default address-model    : 64-bit (cached)
    - default architecture     : x86 (cached)
    - symlinks supported       : yes (cached)
...patience...
...patience...
...patience...
...found 5085 targets...
...updating 4 targets...
quickbook.quickbook-to-boostbook ../../../bin.v2/libs/geometry/doc/gcc-9/debug/python-2.7/threading-multi/visibility-hidden/geometry.xml
Generating output file: ../../../bin.v2/libs/geometry/doc/gcc-9/debug/python-2.7/threading-multi/visibility-hidden/geometry.xml
xslt-xsltproc ../../../bin.v2/libs/geometry/doc/gcc-9/debug/python-2.7/threading-multi/visibility-hidden/geometry.docbook
xslt-xsltproc-dir html/geometry_HTML.manifest
...updated 4 targets...
```

Finally, load `libs/geometry/doc/html/index.html` and you will have all generated content linked from `libs/geometry/doc/html/geometry/*`, clickable and accessible.

digu-007 commented:
@mloskot Can you tell me how to add `boostbook` to `user-config.jam`. While running `make_qbk.py` only 3 targets are getting updated, I think `boostbook` is causing that bug. Although `HTML` files are built now and I can check all my work!

mloskot commented:
@digu-007 
> Can you tell me how to add boostbook to user-config.jam

Well, have you tried to follow http://www.boost.org/doc/libs/release/doc/html/quickbook/install.html ? 

In https://github.com/boostorg/geometry/pull/667#issuecomment-584652584 you said you followed [Generating-and-Improving-Documentation](https://github.com/boostorg/geometry/wiki/Generating-and-Improving-Documentation). Am I missing anything?

> While running make_qbk.py only 3 targets are getting updated

Yes, because your setup is missing the `user-config.jam`.

digu-007 commented:
@mloskot  Yes, I did setup `user-config.jam` according to [this](https://www.boost.org/doc/libs/1_72_0/doc/html/quickbook/install.html). 

Here have a look:
```
using xsltproc ;

using boostbook
    : /usr/share/xml/docbook/stylesheet/nwalsh
    : /usr/share/xml/docbook/schema/dtd/4.2
    ;

# Remove this line if you're not using doxygen
using doxygen ;

using quickbook
    : /usr/local/bin/quickbook
    ;
```

mloskot commented:
Then your question "Can you tell me how to add boostbook to user-config.jam" in https://github.com/boostorg/geometry/pull/667#issuecomment-585212491 is confusing 

Anyway, here is mine

```
$ sudo apt-get install xsltproc docbook-xsl docbook-xml libxslt1.1
```

```
$ cat ~/user-config.jam
using xsltproc ;

using boostbook
    : /usr/share/xml/docbook/stylesheet/nwalsh
    : /usr/share/xml/docbook/schema/dtd/4.2
    ;

using doxygen ;
```

If you are still not getting the HTML generated, try to clean the stuff:

```
$ cd libs/geometry/doc/
$ python3 make_qbk.py --release-build   # this will clean, w/o running b2
$ python3 make_qbk.py                   # this will double-clean and run b2
```

You may see `...updated 3 targets...` or `...updated 4 targets...` depending what `b2` finds out of date (e.g. `bin.v2` folder in Boost root directory has been removed).

However, important is if you see this lines at the end of the `python make_qbk.py` output, that is building of the HTML:

```console
xslt-xsltproc ../../../bin.v2/libs/geometry/doc/gcc-9/debug/python-2.7/threading-multi/visibility-hidden/geometry.docbook
xslt-xsltproc-dir html/geometry_HTML.manifest
```

digu-007 commented:
> Anyway, here is mine
> 
> ```
> $ sudo apt-get install xsltproc docbook-xsl docbook-xml libxslt1.1
> ```
Oh this is the command for `boostbook`, sorry then, I have indeed installed that. 

> You may see `...updated 3 targets...`

Yes I'm getting that, and `HTML` are generated fine.

```
quickbook.quickbook-to-boostbook ../../../bin.v2/libs/geometry/doc/gcc-7/debug/python-3.7/threading-multi/visibility-hidden/geometry.xml
Generating output file: ../../../bin.v2/libs/geometry/doc/gcc-7/debug/python-3.7/threading-multi/visibility-hidden/geometry.xml
xslt-xsltproc ../../../bin.v2/libs/geometry/doc/gcc-7/debug/python-3.7/threading-multi/visibility-hidden/geometry.docbook
xslt-xsltproc-dir html/geometry_HTML.manifest
...updated 3 targets...
```
So I guess, everything is fine and I will work on final integration of the example codes and respective images to the frontend.

mloskot commented:
> > ```
> > $ sudo apt-get install xsltproc docbook-xsl docbook-xml libxslt1.1
> > ```
> 
> Oh this is the command for `boostbook`, sorry then, I have indeed installed that.

That `apt-get` installs DocBook XML/XSL files required by BoostBook, as [BoostBook documentation format is an extension of DocBook](https://www.boost.org/doc/libs/1_72_0/doc/html/boostbook.html).

digu-007 commented:
I finally linked all the examples and images to the frontend. I verified all the results by locally testing.
@mloskot  Kindly, please check.

awulkiew commented:
Btw, AFAIR instead of running `make_qbk.py` manually you can run `b2 doc/`. This way the b2 will install everything for you.

mloskot commented:
> This way the b2 will install everything for you.

I think it's a bit too optimistic 😄 

The `b2` will build `doxygen_xml2qbk` and run the `make_qbk.py` workflow.
It will not install everything though, i.e.  https://github.com/boostorg/geometry/blob/5226efc76f9e49cd67617a286200524e3fb3f4fd/doc/readme.txt#L16-L17

awulkiew commented:
@mloskot yes, I should've written "build the prerequisites". ;)

vissarion commented on doc/src/examples/algorithms/covered_by.cpp:
```diff
@@ -0,0 +1,70 @@
+// Boost.Geometry (aka GGL, Generic Geometry Library)
+// QuickBook Example
+
+// Copyright (c) 2020-2021 Digvijay Janartha, Hamirpur, India.
```
we usually only include the running year, so 2020 is enough

digu-007 commented:
@vissarion I fixed years in all the files.
Kindly review [it](https://github.com/boostorg/geometry/pull/667/commits/e41091ea71072f57d4288278de6c7173a80c16f9).

awulkiew commented:
Thanks for the PR!

digu-007 commented:
@mloskot @vissarion @awulkiew @barendgehrels 
Thanks for assisting and building the documentation! :)

mloskot commented:
FYI, in 53a6ada73c09b3250be66419e797fa2c81134687 I committed quick fix for this issue that sneaked in

```
Generating output file: /root/build/boost/bin.v2/libs/geometry/doc/gcc-5.4.0/debug/python-2.7/threading-multi/visibility-hidden/geometry.xml
/root/project/libs/geometry/doc/src/examples/algorithms/touches_two_geometries.cpp:11: error: Template Redefinition: touches
/root/project/libs/geometry/doc/src/examples/algorithms/touches_two_geometries.cpp:49: error: Template Redefinition: touches_output
```

It broke global Boost docs build, so I had to react quick.

digu-007 commented:
> FYI, in [53a6ada](https://github.com/boostorg/geometry/commit/53a6ada73c09b3250be66419e797fa2c81134687) I committed quick fix for this issue that sneaked in

@mloskot Thank you for solving the bug! I remember that bug occurred during testing, I don't know how I lost track of it before the final commit.
